src/main/java/uk/gov/hmcts/darts/courthouse/service/CourthouseServiceImpl.java

```diff
@@ -121,6 +135,150 @@ public CourthouseEntity addCourtHouse(Courthouse courthouse) {
         return courthouseRepository.saveAndFlush(mappedEntity);
     }
 
+    private final SecurityGroupIdMapper securityGroupIdMapper;
+
+    @Override
+    public ExtendedCourthousePost mapFromEntitiesToExtendedCourthousePost(CourthousePost courthousePost) {
+        final SecurityRoleEntity securityRoleTranscriber = getSecurityRoleByRoleName("TRANSCRIBER");
+        final SecurityRoleEntity securityRoleRequester = getSecurityRoleByRoleName("REQUESTER");
+        final SecurityRoleEntity securityRoleApprover = getSecurityRoleByRoleName("APPROVER");
+
+        //check duplicate names
+        checkCourthouseNameIsUnique(courthousePost);
+        checkCourthouseDisplayNameIsUnique(courthousePost);
+
+        List<Integer> securityGroupIds = (courthousePost.getSecurityGroupIds() == null) ? new ArrayList<>() : courthousePost.getSecurityGroupIds();
+        courthousePost.securityGroupIds(securityGroupIds);
+
+        //for each security_group get role_ids
+        List<Integer> securityRoleIds = new ArrayList<>();
+        for (Integer secGrpIds : securityGroupIds) {
+            Optional<SecurityGroupEntity> securityGroupEntity = securityGroupRepository.findById(secGrpIds);
+            securityGroupEntity.ifPresent(groupEntity -> securityRoleIds.add(groupEntity.getSecurityRoleEntity().getId()));
+        }
+
+        //ensure if security roles exist, TRANSCRIBER role also exists
+        if (securityRoleTranscriber != null && (!securityRoleIds.isEmpty() && !securityRoleIds.contains(securityRoleTranscriber.getId()))) {
+                throw new DartsApiException(AuthorisationError.BAD_REQUEST_ONLY_TRANSCIBER_ROLES_MAY_BE_ASSIGNED);
+
+        }
```

Comment:
As this code is concerned specifically with validation, consider splitting it out into a separate private method, `validate(CourthousePost courthousePost);`

This would make this method a bit more readable, as validation would be abstracted away.

Reply:
securityRoleIds are used later in the code so a private method getSecurityRoleIds(courthousePost, securityRoleTranscriber) was created

Reply:
I have addressed this